include/dali/core/stream.h

```diff
@@ -0,0 +1,204 @@
+   */
+  inline void ReadBytes(void *buf, size_t bytes) {
+    char *b = static_cast<char *>(buf);
+    while (bytes) {
+      ssize_t n = Read(b, bytes);
+      if (n == 0)
+        throw EndOfStream();
+      if (n < 0)
+        throw std::runtime_error("An error occurred while reading data.");
+      b += n;
+      assert(static_cast<size_t>(n) <= bytes);
+      bytes -= n;
+    }
+  }
+
+  /**
+   * @brief Reads one object of given type from the stream
+   *
+   * @tparam T  the type of the object to read; should be trivially copyable or otherwise
+   *            safe to be overwritten with memcpy or similar.
+   */
+  template <typename T>
+  inline T ReadOne() {
+    T t;
+    ReadAll(&t, 1);
+    return t;
+  }
+
+  /**
+   * @brief Reads `count` instances of type T from the stream to the provided buffer
+   *
+   * If the function cannot read the requested number of objects, an exception is thrown
+   *
+   * @tparam T    the type of the object to read; should be trivially copyable or otherwise
+   *              safe to be overwritten with memcpy or similar.
+   * @param buf   the output buffer
+   * @param count the number of objects to read
+   */
+  template <typename T>
+  inline void ReadAll(T *buf, size_t count) {
+    ReadBytes(buf, sizeof(T) * count);
+  }
+
+  /**
+   * @brief Skips `count` objects in the stream
+   *
+   * Skips over the length of `count` objects of given type (by default char,
+   * because sizeof(char) == 1).
+   *
+   * NOTE: Negative skips are allowed.
+   *
+   * @tparam T type of the object to skip; defaults to `char`
+   * @param count the number of objects to skip
+   */
+  template <typename T = char>
+  void Skip(ssize_t count = 1) {
+    SeekRead(count * sizeof(T), SEEK_CUR);
+  }
+
+  /**
+   * @brief Reads data from the stream and advances the read pointer; partial reads are allowed.
+   *
+   * A valid implementation of this function reads up to `bytes` bytes from the stream and
+   * stores them in `buf`. If the function cannot read all of the requested bytes due to
+   * end-of-file, it shall read all it can and return the number of bytes actually read.
+   * When reading from a regular file and the file pointer is already at the end, the function
+   * shall return 0.
+   *
+   * This function does not throw EndOfStream.
+   *
+   * @param buf       the output buffer
+   * @param bytes     the number of bytes to read
+   * @return size _t  the number of bytes actually read or
+   *                  0 in case of end-of-stream condition
+   */
+  virtual size_t Read(void *buf, size_t bytes) = 0;
+
+  /**
+   * @brief Moves the read pointer in the stream.
+   *
+   * If the new pointer would be out of range, the function may either clamp it to a valid range
+   * or throw an error.
+   *
+   * @param pos     the offset to move
+   * @param whence  the beginning - SEEK_SET, SEEK_CUR or SEEK_END
+   */
+  virtual void SeekRead(ptrdiff_t pos, int whence = SEEK_SET) = 0;
```

Comment:
This is `SeekRead` and not just `Seek` for two reasons:
1. To avoid name collision when we have InputStream and OutputStream and a class that inherits both
2. To emphasize that SeekRead and SeekWrite may be independent (although they don't have to - fseek will move both).